### DISCUSSION TITLE
Batch join defined number of images

### DISCUSSION THREAD

**Discussion by GITUPGITDOWN123** [ImageMagick/ImageMagick#1998] [General]
Hi. I'm new to this and looking for some guidance if possible. What I'd like to be able to do is take a folder of 300 images and batch join two images side by side. That is, image 1 and 2 get joined side by side, and then image 3 and 4 get joined side by side and so on. The result would be 150 images. Is this possible and how best do I approach this? Thank you for any help here.

**Comment by fmw42** [comment]
You would have to write a script loop. Your images would need to be listed in your directory in alphabetic order to take them in sequence. For example image_10 comes before image_1. So you need to have leading zeroes in your numbers, e.g. image_001, image_002 ...

What is your ImageMagick version and platform?  Scripting is quite different for Windows vs Linux/Mac.

**Reply by GITUPGITDOWN123** [reply]
All of my images are screenshots that have been adjusted with Apple Automator to have date and time identifiers added to the file name. For example, I would want to join the first two file and then the second two files:

05-06-2020-20-26-46-screenshot
05-06-2020-20-26-50-screenshot
05-06-2020-20-30-01-screenshot
05-06-2020-20-30-05-screenshot

I'm on a Mac/Catalina 10.5 and haven't downloaded ImageMagick just yet. How difficult is it to create a script loop? Thank you.

**Comment by fmw42** [comment]
How do you want to name the output images and where do you want to put them?

**Reply by GITUPGITDOWN123** [reply]
I think it would make sense if the merged images were placed in a new folder on the desktop and named sequentially by the time generated.

For example, 
05-06-2020-20-26-46-screenshot
05-06-2020-20-26-50-screenshot
could become
07-07-2020-09-07-50-merged

**Comment by fmw42** [comment]
A bash shell script would look something like the following (that works on my Mac). I have images in folder test1 and an empty folder test2, both on my desktop.

```
cd
cd desktop/test1
arr=(`ls`)
num=${#arr[*]}
numm1=$((num-1))
for ((i=0; i<numm1; i=i+2)); do
j=$((i+1))
dt=`date '+%d-%m-%Y-%H-%M-%S'`
sleep 1
convert "${arr[$i]}" "${arr[$j]}" +append ../test2/${dt}-merged.jpg
done

```

**Comment by GITUPGITDOWN123** [comment]
Oh that was really nice of you and I wasn't expecting such great help. It sounds like it can be done which is great news. I'll have to play around some since I can't seem to implement your code properly. What I've tried doing in the inclosed screen shot is to create an Apple Automator workflow that runs the script when prompted. It seems to be having an issue with line 9 'convert' command.
<img width="1421" alt="Screen Shot 2020-05-07 at 10 44 20 AM" src="https://user-images.githubusercontent.com/64971416/81327504-39d0f380-9050-11ea-9d86-c1ea0bd42ace.png">

**Comment by fmw42** [comment]
You have to install ImageMagick.

If you have imagemagick 7, then change convert to magick and try that.

Fred

**Reply by GITUPGITDOWN123** [reply]
Looks installed to me. I wonder what I'm doing wrong.
<img width="1551" alt="Screen Shot 2020-05-07 at 11 31 55 AM" src="https://user-images.githubusercontent.com/64971416/81331434-743d8f00-9056-11ea-823f-eb399afaf268.png">

**Comment by fmw42** [comment]
Does it work if not in an AppleScript. Simply copy and paste in bash terminal (be sure it is bash).  Do you have folders test1 and test2 on your desktop?

Fred

**Comment by fmw42** [comment]
Did you put the path to ImageMagick magick in your $PATH environment variable or bash_profile file? If not then you need to put the full path to magick in the command.

Fred

**Comment by GITUPGITDOWN123** [comment]
Thank you. I got it to work! It was two things:
1) I had to put in the path as you said:
/usr/local/Cellar/imagemagick/7.0.10-10/bin/magick

and

2) It didn't seem to like the file names used in Mac's default screenshot tool aka "Screen Shot 2020-05-07 at 10.31.45 AM". Once I changed the starting file names to A, B, C, and D it worked.

Thank you so much Fred. This is a great step forward and I'm going to play around some more now. Really appreciate it.

**Comment by fmw42** [comment]
Be sure to quote filenames that have spaces in them. I try to never use spaces in file names.

**Comment by fmw42** [comment]
Also see the Pointers for use on my home page.

**Comment by GITUPGITDOWN123** [comment]
Will do. Have a great day.

**Comment by GITUPGITDOWN123** [comment]
One quick thing I meant to ask. If it's easy, can you show me where/how to adjust code to allow say three images to be joined vs the two images in the above example. Once I see the logic in increasing the number of images I should be able to tweak it as needed. 

cd
cd desktop/test1
arr=(`ls`)
num=${#arr[*]}
numm1=$((num-1))
for ((i=0; i<numm1; i=i+2)); do
j=$((i+1))
dt=`date '+%d-%m-%Y-%H-%M-%S'`
sleep 1
convert "${arr[$i]}" "${arr[$j]}" +append ../test2/${dt}-merged.jpg
done

**Comment by fmw42** [comment]
Try (untested)

cd
```
cd desktop/test1
arr=(ls)
num=${#arr[*]}
numm1=$((num-1))
for ((i=0; i<numm1; i=i+3)); do
j=$((i+1))
k=$((i+2))
dt=date '+%d-%m-%Y-%H-%M-%S'
sleep 1
convert "${arr[$i]}" "${arr[$j]}" "${arr[$k]}" +append ../test2/${dt}-merged.jpg
done

```

**Comment by GITUPGITDOWN123** [comment]
Oh I get it. Makes sense and works. Thanks.
